src/ui/public/kfetch/index.js

```diff
@@ -57,17 +67,28 @@ export async function kfetch(fetchOptions, kibanaOptions) {
     query,
   });
 
-  const res = await fetch(fullUrl, combinedFetchOptions);
+  const fetching = new Promise(async (resolve, reject) => {
+    const res = await fetch(fullUrl, combinedFetchOptions);
 
-  if (!res.ok) {
-    let body;
-    try {
-      body = await res.json();
-    } catch (err) {
-      // ignore error, may not be able to get body for response that is not ok
+    if (!res.ok) {
+      let body;
+      try {
+        body = await res.json();
+      } catch (err) {
+        // ignore error, may not be able to get body for response that is not ok
+      }
+      reject(new FetchError(res, body));
     }
-    throw new FetchError(res, body);
+
+    resolve(res.json());
+  });
+
+  if (isAbortable) {
+    return {
+      fetching,
+      abort,
+    };
   }
```

Comment:
I originally considered this approach but I had a few concerns about monkey-patching a native API:

1. If a native `abort` method is introduced to promises at some point in the future then we'll need to ensure it either has identical functionality, or make a decision as to whether to cut over to the official method and ensure BWC or to keep the monkey-patched method and have unexpected behavior.

2. Monkey-patching has the potential to confuse consumers as well since this method isn't in the spec, prompting questions like "Is this a native promise?" and "What other methods are attached to this promise?" It requires the user to understand the implementation in order to have confidence about what's changed about this promise.

These concerns are mostly based on principle so if you think it's more pragmatic to just monkey-patch the promise, let me know your thoughts because I can definitely be persuaded! OTOH if you think these concerns are worthy then I think creating a secondary API makes sense, because we're highlighting the fact that the consumer is consciously opting into secondary behavior. And realistically, do you think it will be used very often? Do we have many use cases for aborting a fetch?